### PR TITLE
Feature ticket type methods

### LINKED ISSUE #353: Add endpoint to change TicketType.
This is to extend the early bird deadline beyond stufi.

### PR DESCRIPTION
**Pull request by MatthijsKok**
Fixes #353

### CODE REVIEW

File: src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java [hunk 1]
```diff
@@ -119,13 +119,32 @@ public TicketRestController(TicketService ticketService, OrderService orderServi
...
         TicketType ticketType = ticketService.addTicketType(type);
-
         return createResponseEntity(HttpStatus.CREATED, "TicketType successfully added", ticketType);
     }
 
     @PreAuthorize("hasRole('ADMIN')")
+    @GetMapping("/types")
+    public ResponseEntity<?> readTicketTypes() {
+        Collection<TicketType> ticketTypes = ticketService.getAllTicketTypes();
+        return createResponseEntity(HttpStatus.OK, "TicketTypes successfully retrieved.", ticketTypes);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java]: You're getting a collection here, so just return those. No need for a wrapping response entity

File: src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java [hunk 2]
```diff
@@ -119,13 +119,32 @@ public TicketRestController(TicketService ticketService, OrderService orderServi
...
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PutMapping("/types")
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java]: This mapping implies replacing all TicketTypes

File: src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java [hunk 3]
```diff
@@ -119,13 +119,32 @@ public TicketRestController(TicketService ticketService, OrderService orderServi
...
+    public ResponseEntity<?> updateTicketType(@RequestBody @Validated TicketType type) {
+        TicketType ticketType = ticketService.updateTicketType(type);
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java]: Which type are you actually updating here?

File: src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java [hunk 4]
```diff
@@ -119,13 +119,32 @@ public TicketRestController(TicketService ticketService, OrderService orderServi
...
+        return createResponseEntity(HttpStatus.OK, "TicketType successfully updated.", ticketType);
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @DeleteMapping("/types")
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java]: This implies deleting all TicketTypes. It should be a DELETE on `/types/{id}` or something like that.

File: src/test/java/ch/wisv/areafiftylan/products/service/TicketServiceTest.java [hunk 5]
```diff
@@ -650,6 +652,62 @@ public void getAllTicketTypesTwoExtra() {
...
+        ticketTypeRepository.save(ticketType2);
+        ticketType = ticketService.updateTicketType(ticketType.getId(), ticketType);
+        assertTrue(ticketType.isBuyable());
+    }
+
+    @Test
+    public void updateTicketTypeTestNoneThere() {
+        TicketType ticketType = new TicketType("type1", "text", 5F, 0, LocalDateTime.now(), false);
+        ticketType = ticketService.updateTicketType(ticketType.getId(), ticketType);
+        assertFalse(ticketType.isBuyable());
```
Comment (skamoen) [inline src/test/java/ch/wisv/areafiftylan/products/service/TicketServiceTest.java]: I dont think this is the kind of assertion you want for this test

File: src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java [hunk 6]
```diff
@@ -119,13 +119,35 @@ public TicketRestController(TicketService ticketService, OrderService orderServi
...
+    @GetMapping("/types")
+    public Collection<TicketType> readTicketTypes() {
+        return ticketService.getAllTicketTypes();
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PutMapping("/types/{typeId}")
+    public ResponseEntity<?> updateTicketType(@PathVariable Long typeId, @RequestBody @Validated TicketType type) {
+        TicketType ticketType = ticketService.updateTicketType(typeId, type);
+        if (ticketType == null) {
```
Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java]: This is a weird check for a controller. When would this be null exactly? You already made the call to the service here...

Comment (MatthijsKok) [inline src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java]: The service returns null if the save was not done successfully (when typeId doesn't exist for example)

Comment (skamoen) [inline src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java]: Then I'd suggest to throw some exceptions in the service for errors like that.

Comment (MatthijsKok) [inline src/main/java/ch/wisv/areafiftylan/products/controller/TicketRestController.java]: Good point. Will do

**Pull request merged**